Thermoelasticity with reference temperatures

stvdwtt opened the pull request:
This PR does the following things:
- Adds a new state that tracks if the cell has been melted
- Adds functionality to adjust the state if the cell goes above a certain temperature
- Changes the thermal expansion to be vs a reference temperature that depends on whether the cell has melted
- Tests

stvdwtt commented:
The [CI](https://cloud.cees.ornl.gov/jenkins-ci/blue/organizations/jenkins/adamantine/detail/adamantine/332/pipeline) failed, I'm going to look into it this afternoon. Oddly, it's passing on my machine. It could be an inconsistency between test input files in the repo and in my build directory.

Rombur commented:
I don't know if you know but if you click on the `Tests` button on the top, you can see the error messages from the tests (see [here](https://cloud.cees.ornl.gov/jenkins-ci/blue/organizations/jenkins/adamantine/detail/adamantine/332/tests)). I don't think the problem is the input file. Are you using the latest version of the Docker image?

stvdwtt commented:
I'm using the lastest version of `rombur/adamantine-stack:latest`, do you know if that has diverged from `rombur/adamantine:latest`?

Rombur commented:
We use `rombur/adamantine-stack:latest` for the CI not `rombur/adamantine:latest`. `rombut/adamantine:latest` is automatically created when we merge a PR.

stvdwtt commented:
I added a few more changes to address the errors in the CI. 

- Fixes a mistake in how I handled hp elements
- Changes the PostProcess object to a unique pointer to address cases where mechanics is and is not included
- Changes Jenkins input to compile the "CPU" test with deal.ii weak forms
- Adds a check that deal.ii weak forms is linked if the input says that mechanical physics should be active (addresses #210)

Link to the new CI: https://cloud.cees.ornl.gov/jenkins-ci/blue/organizations/jenkins/adamantine/detail/adamantine/338/pipeline/

stvdwtt commented:
@Rombur, the CI passed and this is ready for review.

Rombur commented on application/adamantine.hh:
```diff
@@ -882,27 +898,56 @@ run(MPI_Comm const &communicator, boost::property_tree::ptree const &database,
       communicator, post_processor_database,
       thermal_physics->get_dof_handler());
 
+#endif
```
Let's not do this because we still need to add  a case where only the mechanical simulation is enabled. It's also harder to read because of the mix of `#ifdef` and `if`. Instead let's do sth like this:
```c++
thermal_output = post_processor_database.get("thermal_output", false);
mechanical_output = post_processor_database.get("mechanical_output", false);
if ((thermal_output) && (mechanical_output))
  {
     post_processor = std::make_unique<adamantine::PostProcessor<dim>>(
        communicator, post_processor_database,
        thermal_physics->get_dof_handler(),
        mechanical_physics->get_dof_handler());
  }
else
 {
    post_processor = std::make_unique<adamantine::PostProcessor<dim>>(
        communicator, post_processor_database,
        thermal_output ? thermal_physics->get_dof_handler() : mechanical_physics->get_dof_handler());
 }

```

Rombur commented on application/adamantine.hh:
```diff
@@ -1059,6 +1109,17 @@ run(MPI_Comm const &communicator, boost::property_tree::ptree const &database,
+    // thermomechanical solves.
+    if (use_thermal_physics && use_mechanical_physics)
+    {
+      thermal_physics->mark_cells_above_temperature(material_reference_temps[1],
```
Maybe call this `mark_has_melted`? Above temperature is not very obvious which temperature we are talking about.

stvdwtt replied:
Sounds good -- that fits with everything else better.

Rombur commented on source/MechanicalOperator.cc:
```diff
@@ -84,10 +84,12 @@ template <int dim, typename MemorySpaceType>
     dealii::LA::distributed::Vector<double, dealii::MemorySpace::Host> const
-        &temperature)
+        &temperature,
+    std::vector<double> const &has_melted_indicator)
```
Why use a `double`? I think a `bool` would be more appropriate.

stvdwtt replied:
I agree that a `bool` would fit better, but my understanding is that it has to be a `double` to get packed into the state. It's a bit more indirect, but if you think it would be better, I can store the vector as `bool`s and then convert to `double`s element by element as I pack and unpack the state.

Rombur commented on source/MechanicalOperator.cc:
```diff
@@ -188,12 +190,35 @@ void MechanicalOperator<dim, MemorySpaceType>::assemble_system()
+          {
+            is_unmelted_substrate = true;
+          }
+          // const bool is_unmelted_substrate = cell->user_flag_set();
```
I guess we don't need this anymore?

stvdwtt replied:
Right, we don't use the `user_flag`s anymore

Rombur commented on source/MechanicalOperator.cc:
```diff
@@ -188,12 +190,35 @@ void MechanicalOperator<dim, MemorySpaceType>::assemble_system()
+
+          // FIXME: Needs new implementation
+          bool is_unmelted_substrate = false;
+          if (_has_melted_indicator.at(cell->active_cell_index()) < 0.5)
```
Why are using `at`? It is significantly slower than `[]`

stvdwtt replied:
I'm switching it to `[]`, that was to help me catch some indexing issues.

Rombur commented on source/MechanicalOperator.hh:
```diff
@@ -102,11 +103,13 @@ private:
-   * solve a themo-mechanical problem. Otherwise, we solve a mechanical only
-   * problem.
+   * List of initial temperatures of the material. If the length of the vector
+   * is nonzero, we solve a themo-mechanical problem. Otherwise, we solve a
```
```suggestion
   * is nonzero, we solve a thermo-mechanical problem. Otherwise, we solve a
```

Rombur commented on source/MechanicalOperator.cc:
```diff
@@ -188,12 +190,35 @@ void MechanicalOperator<dim, MemorySpaceType>::assemble_system()
+          else
+          {
+            reference_temperature =
+                _reference_temperatures.at(cell->material_id() + 1);
```
Why the `+1`?

stvdwtt replied:
The reference temperature vector looks like: {initial condition temperature, solidus material 0, solidus material 1, ...}

Rombur replied:
What about we switch the order? {solidus_material_0, solidus material, ..., initial condition temperature} Then when you call `_reference_temperature[cell->material_id()]`, you get what you would expect.

stvdwtt replied:
I like that idea a lot

Rombur commented on source/MechanicalOperator.hh:
```diff
@@ -141,6 +144,13 @@ private:
+   * value is 0 for material that has not yet melted and 1 for material that has
+   * melted. Due to how the state is transferred for remeshing, this has to be
+   * stored as a double.
+   */
```
You can just use a `double` when you refine and then convert back to a `bool`.

stvdwtt replied:
That makes sense, I'll make that change.

Rombur commented on source/ThermalPhysics.templates.hh:
```diff
@@ -567,6 +569,62 @@ void ThermalPhysics<dim, fe_degree, MemorySpaceType,
+      }
+    }
+
+  } // namespace adamantine
```
I don't think this comment should be here.

stvdwtt replied:
Good catch, removed.

Rombur commented on source/validate_input_database.cc:
```diff
@@ -24,6 +24,13 @@ void validate_input_database(boost::property_tree::ptree &database)
                "Error: both thermal and mechanical physics are disabled");
+  if (use_mechanical_physics)
+  {
+#ifndef ADAMANTINE_WITH_DEALII_WEAK_FORMS
```
Move the `#ifndef` about the `if`

stvdwtt commented on ci/jenkins_config:
```diff
@@ -56,8 +56,10 @@ pipeline {
                     cmake \
                     -D CMAKE_BUILD_TYPE=Debug \
                     -D ADAMANTINE_ENABLE_TESTS=ON \
+                    -D ADAMANTINE_ENABLE_DEALII_WEAK_FORMS=ON \
```
Oh, I see. In that case, what should I be doing for the thermoelastic integration test? For obvious reasons it fails without deal.II weak forms. Should the test be catching an exception in the case that deal.II weak forms isn't linked? I'm a little puzzled how the mechancial_operator and mechanical_physics tests were passing.

Rombur replied:
The tests are not run if we don't have dealii-weak_forms. See [here](https://github.com/adamantine-sim/adamantine/blob/master/tests/CMakeLists.txt#L69-L73)

stvdwtt replied:
Ah, I see. Thanks! I'll move that test to the WEAK_FORMS_UNIT_TESTS and revert the CPU cmake options.

stvdwtt commented on application/adamantine.hh:
```diff
@@ -882,27 +898,56 @@ run(MPI_Comm const &communicator, boost::property_tree::ptree const &database,
     // PropertyTreeInput discretization.mechanical.fe_degree
     unsigned int const fe_degree =
-        discretization_database.get<unsigned int>("mechanial.fe_degree");
+        discretization_database.get<unsigned int>("mechanical.fe_degree");
```
I don't think it did. We didn't have any integration tests with mechanics.

stvdwtt commented on application/adamantine.hh:
```diff
@@ -414,13 +418,17 @@ void refine_and_transfer(
         cell_data[n_material_states + 1] =
             thermal_physics->get_deposition_sin(activated_cell_id);
+        cell_data[n_material_states + direction_data_size] =
+            thermal_physics->get_has_melted_indicator(activated_cell_id);
```
Good idea, I'll make that change.

Rombur commented on application/adamantine.hh:
```diff
@@ -882,26 +907,59 @@ run(MPI_Comm const &communicator, boost::property_tree::ptree const &database,
+  material_reference_temps.push_back(initial_temperature);
+
+// Create MechanicalPhysics
 #ifdef ADAMANTINE_WITH_DEALII_WEAK_FORMS
```
This needs to be moved down after `mechanical_physics` so that `mechanical_physics` always exist because now we use it lower.

Rombur commented on application/adamantine.hh:
```diff
@@ -518,6 +532,14 @@ void refine_and_transfer(
+                            [n_material_states + direction_data_size] > 0.5)
+          has_melted.push_back(true);
+        else
+
```
There is another extra blank line

stvdwtt commented:
I added the dummy `MechanicalPhysics` and I believe have addressed all of the review comments.

[CI](https://cloud.cees.ornl.gov/jenkins-ci/blue/organizations/jenkins/adamantine/detail/adamantine/343/pipeline)
